File: ios/Splitty/Features/Activity/ActivityViewModel.swift
import Foundation
import Observation

/// VM вкладки «Уведомления»: лента операций всех групп с пагинацией offset/limit.
/// Первая страница читается через `DataRepo` (офлайн-кеш): кеш показывается
/// мгновенно, сетевая ошибка при наличии кеша — без алерта. Следующие
/// страницы не кешируются (офлайн лента ограничена первой страницей).
@MainActor
@Observable
final class ActivityViewModel {
    /// Состояние загрузки ленты.
    enum LoadState {
        case idle
        case loading
        case loaded
        case failed(String)
    }

    private static let pageSize = 30
    /// За сколько строк до конца списка начинать подгрузку следующей страницы.
    private static let prefetchThreshold = 5

    var state: LoadState = .idle
    var items: [ActivityItem] = []
    /// Закреплённые карточки приглашений над лентой.
    var invites: [InviteCard] = []
    /// Непрочитанное: pending-приглашения + непрочитанные added + новые события.
    private(set) var unreadCount = 0
    /// Время формирования последнего ответа — его же отправляем при отметке
    /// прочитанного, чтобы не погасить то, что пришло позже.
    private var seenThrough: Date?
    /// true — показан офлайн-кеш (сеть недоступна), не свежие данные.
    private(set) var isFromCache = false
    /// Когда данные последний раз пришли С СЕРВЕРА — для подписи о свежести.
    private(set) var lastUpdatedAt: Date?
    /// Ошибка обновления/подгрузки, когда лента уже показана (для alert).
    var errorMessage: String?
    private(set) var isLoadingMore = false
    private var hasMore = true

    /// Первичная загрузка; при повторном появлении вкладки — тихое обновление.
    func load(repo: DataRepo) async {
        switch state {
        case .loading:
            return
        case .loaded:
            await reload(repo: repo)
        case .idle, .failed:
            state = .loading
            await reload(repo: repo)
        }
    }

    /// Отметить прочитанным всё, что было в последнем ответе.
    ///
    /// Отправляем СЕРВЕРНЫЙ `seenThrough` из ответа, а не текущее время: между
    /// ответом и этим вызовом мог прийти новый расход, и «сейчас» погасило бы
    /// его, так и не показав человеку.
    func markSeen(session: SessionStore) async {
        // Публикуем счётчик, только когда за ним стоит настоящий ответ (сетевой
        // или из кеша). Лента могла не загрузиться — нет сети, ошибка сервера,
        // отменённый .task, — и тогда unreadCount ноль просто потому, что его
        // никто не считал: обнулив бейдж, приложение соврало бы о пустых
        // входящих, а отметку прочитанного при этом никто не отправлял.
        guard let through = seenThrough else { return }
        // Счётчик из последнего ответа — источник правды для бейджа, и записать
        // его надо ДО раннего выхода: если отметку уже поставили с другого
        // устройства, unreadCount придёт нулём, отмечать будет нечего, а бейдж
        // так и висел бы до следующего возврата из фона.
        session.unreadNotifications = unreadCount
        guard unreadCount > 0 else { return }
        do {
            try await session.api.markNotificationsSeen(through: through)
            unreadCount = invites.filter { $0.status == .pending }.count
            session.unreadNotifications = unreadCount
        } catch {
            // Не показываем алерт: отметка прочитанного — фоновое действие,
            // человек её не запрашивал явно.
            if error.isTaskCancellation { return }
        }
    }

    /// Принять приглашение вернуться в группу.
    func acceptInvite(_ card: InviteCard, session: SessionStore) async {
        await actOnInvite(card, session: session) {
            try await session.api.acceptInvite(roomId: card.roomId)
        }
    }

    /// Отклонить приглашение.
    func declineInvite(_ card: InviteCard, session: SessionStore) async {
        await actOnInvite(card, session: session) {
            try await session.api.declineInvite(roomId: card.roomId)
        }
    }

    /// Выйти из группы прямо с карточки «вас добавили».
    ///
    /// Кнопка обязана быть здесь: человека добавили, не спросив, и если
    /// единственное действие — «Открыть», отказаться можно только разыскав
    /// настройки группы.
    func leaveFromCard(_ card: InviteCard, session: SessionStore) async {
        await actOnInvite(card, session: session) {
            try await session.api.leaveRoom(roomId: card.roomId)
        }
    }

    private func actOnInvite(
        _ card: InviteCard,
        session: SessionStore,
        _ action: () async throws -> Void
    ) async {
        do {
            try await action()
            invites.removeAll { $0.roomId == card.roomId }
            // Данные комнат изменились — списки групп и друзей перечитаются.
            session.noteDataChanged()
        } catch {
            if error.isTaskCancellation { return }
            // Тот же текст, что в настройках группы: 409 сюда приходит только
            // с кнопки «Выйти», и он обязан объяснять путь наружу.
            errorMessage = leaveErrorText(error)
        }
    }

    /// Pull-to-refresh: перезагрузка первой страницы без скрытия ленты.
    func refresh(repo: DataRepo) async {
        await reload(repo: repo)
    }

    /// Подгружает следующую страницу, когда пользователь долистал до `item`.
    func loadMoreIfNeeded(repo: DataRepo, current item: ActivityItem) async {
        guard case .loaded = state, hasMore, !isLoadingMore else { return }
        guard let index = items.firstIndex(where: { $0.id == item.id }),
              index >= items.count - Self.prefetchThreshold else { return }
        await loadNextPage(repo: repo)
    }

    private func loadNextPage(repo: DataRepo) async {

        isLoadingMore = true
        defer { isLoadingMore = false }
        do {
            let page = try await repo.api.activity(limit: Self.pageSize, offset: items.count)
            // Страховка от дублей при сдвиге offset (новые операции сверху).
            let known = Set(items.map(\.id))
            items += page.filter { !known.contains($0.id) }
            hasMore = page.count == Self.pageSize
        } catch {
            // Отмена .task (строка уехала за экран, ушли с вкладки) — не ошибка,
            // молча выходим без ложного «Нет соединения с сервером».
            if error.isTaskCancellation { return }
            errorMessage = humanErrorText(error)
        }
    }

    private func reload(repo: DataRepo) async {
        do {
            let result = try await repo.notificationFeedFirstPage(limit: Self.pageSize) { [weak self] cached in
                // Кеш мгновенно — только пока в памяти нет более свежих данных.
                guard let self, self.items.isEmpty else { return }
                self.items = cached.items
                self.invites = cached.invites
                self.unreadCount = cached.unreadCount
                self.isFromCache = true
                self.hasMore = false
                self.state = .loaded
            }
            let feed = result.value
            let page = feed.items
            items = page
            invites = feed.invites
            unreadCount = feed.unreadCount
            seenThrough = feed.seenThrough
            isFromCache = result.isFromCache
            if !result.isFromCache { lastUpdatedAt = Date() }
            // Из кеша дальше не листаем (следующие страницы не кешируются) —
            // иначе офлайн-прокрутка до конца ленты давала бы ложный алерт.
            hasMore = result.isFromCache ? false : page.count == Self.pageSize
            state = .loaded
        } catch {
            if error.isTaskCancellation {
                // Отмена посреди первичной загрузки: откатываемся в idle,
                // чтобы следующий .task загрузил ленту заново.
                if case .loading = state { state = .idle }
                return
            }
            if items.isEmpty {
                state = .failed(humanErrorText(error))
            } else {
                errorMessage = humanErrorText(error)
            }
        }
    }
}

File: android/app/src/main/java/com/zagir/splitty/ui/activity/ActivityViewModel.kt
package com.zagir.splitty.ui.activity

import com.zagir.splitty.ui.components.humanErrorText
import com.zagir.splitty.R
import com.zagir.splitty.core.ui.UiText
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.zagir.splitty.core.UiState
import com.zagir.splitty.core.model.ActivityItem
import com.zagir.splitty.core.model.InviteCard
import com.zagir.splitty.core.model.InviteStatus
import com.zagir.splitty.core.network.ApiException
import com.zagir.splitty.core.session.SessionStore
import com.zagir.splitty.data.OutboxSyncer
import com.zagir.splitty.data.SplittyRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

/**
 * VM вкладки «Уведомления»: лента операций всех групп (GET /activity)
 * с пагинацией offset/limit по скроллу. Первичная загрузка и перезагрузка
 * первой страницы после каждой мутации — по [SessionStore.dataVersion].
 * Порт ios/Splitty/Features/Activity/ActivityViewModel.swift.
 *
 * Лента отдаётся экрану как есть: раздел стал входящими, счётчик
 * непрочитанного на сервере считает адресованное вам (`notifiesUser`), и
 * фильтр «Только мои», переключавший ленту между «мне» и «всё подряд»,
 * этому противоречил.
 */
@HiltViewModel
class ActivityViewModel @Inject constructor(
    private val repository: SplittyRepository,
    private val sessionStore: SessionStore,
    private val outboxSyncer: OutboxSyncer,
) : ViewModel() {

    companion object {
        private const val PAGE_SIZE = 30

        /** За сколько строк до конца списка начинать подгрузку следующей страницы. */
        private const val PREFETCH_THRESHOLD = 5
    }

    private val _state = MutableStateFlow<UiState<List<ActivityItem>>>(UiState.Loading)
    val state: StateFlow<UiState<List<ActivityItem>>> = _state.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    private val _isLoadingMore = MutableStateFlow(false)
    val isLoadingMore: StateFlow<Boolean> = _isLoadingMore.asStateFlow()

    /** Ошибка обновления/подгрузки, когда лента уже показана (alert). */
    /** Закреплённые карточки приглашений над лентой. */
    private val _invites = MutableStateFlow<List<InviteCard>>(emptyList())
    val invites: StateFlow<List<InviteCard>> = _invites.asStateFlow()

    /** Время формирования последнего ответа — его же шлём при отметке
     *  прочитанного, чтобы не погасить пришедшее позже. */
    private var seenThrough: java.time.Instant? = null

    /** Счётчик из последнего ответа: нечего гасить — незачем и ходить на сервер. */
    private var unreadCount = 0

    /**
     * Экран на виду. Отметка прочитанного зависит от ДВУХ событий — открытия
     * раздела и прихода ленты, — и первый визит всегда наступает раньше
     * второго: `seenThrough` берётся из ответа, до него отмечать нечего.
     * Поэтому оба пути ведут в [markSeen], а флаг связывает их.
     */
    private var isScreenVisible = false

    private val _errorMessage = MutableStateFlow<UiText?>(null)
    val errorMessage: StateFlow<UiText?> = _errorMessage.asStateFlow()

    /** id текущего пользователя — для позиций «Вы одолжили/должны/получили». */
    val myUserId: StateFlow<Long?> = sessionStore.state
        .map { it?.me?.id }
        .stateIn(viewModelScope, SharingStarted.Eagerly, sessionStore.state.value?.me?.id)

    private var hasMore = true

    /**
     * Сколько строк реально отдал сервер. Именно это, а не размер списка в UI —
     * корректный offset: дубликаты (новые операции сверху сдвигают окно) из
     * списка выбрасываются, и offset по его размеру отставал. Когда сверху
     * появлялась целая страница новых операций, следующая страница приходила
     * полностью дублирующей, размер не менялся, hasMore оставался true — и тот
     * же offset запрашивался бесконечно.
     */
    private var loadedCount = 0

    /** Поколение ленты: подгрузка, стартовавшая до refresh, не должна вернуть старый список. */
    private var generation = 0

    init {
        viewModelScope.launch {
            sessionStore.dataVersion.collect { reloadFirstPage() }
        }
    }

    /** Повтор после ошибки первичной загрузки. */
    fun retry() {
        _state.value = UiState.Loading
        viewModelScope.launch { reloadFirstPage() }
    }

    /** Pull-to-refresh: перезагрузка первой страницы без скрытия ленты. */
    fun refresh() {
        outboxSyncer.syncNow() // pull-to-refresh — один из триггеров досылки outbox
        viewModelScope.launch {
            _isRefreshing.value = true
            try {
                reloadFirstPage()
            } finally {
                _isRefreshing.value = false
            }
        }
    }

    /**
     * Зовётся при показе строки с индексом [index] (snapshotFlow по
     * LazyListState): у конца списка подгружает следующую страницу.
     */
    fun onItemShown(index: Int) {
        val items = (_state.value as? UiState.Content)?.value ?: return
        if (!hasMore || _isLoadingMore.value) return
        if (index < items.size - PREFETCH_THRESHOLD) return
        viewModelScope.launch { loadMore() }
    }

    fun dismissError() {
        _errorMessage.value = null
    }

    /**
     * Раздел открыт — значит человек всё это увидел.
     *
     * Перезагружаем ленту, а не просто отмечаем прочитанное: VM переживает
     * переключение табов (`saveState`/`restoreState`), и на повторном входе
     * `seenThrough` с `unreadCount` остались бы от прошлого визита. Пришедшее
     * в фоне приглашение не показалось бы вовсе, а бейдж, поднятый обновлением
     * счётчика на старте, гасить было бы нечем — `markSeen` уходил бы в no-op.
     * Отметку пошлёт сам [reloadFirstPage], когда приедет ответ.
     */
    fun onScreenVisible() {
        isScreenVisible = true
        viewModelScope.launch { reloadFirstPage() }
    }

    fun onScreenHidden() {
        isScreenVisible = false
    }

    /** Отметить прочитанным всё, что было в последнем ответе. */
    fun markSeen() {
        val through = seenThrough ?: return
        if (unreadCount == 0) return
        viewModelScope.launch {
            // Фоновое действие: сбой не должен ничем мигать пользователю.
            runCatching { repository.markNotificationsSeen(through) }
            // Бейдж гасим только после подтверждённой отметки: иначе непрочитанное
            // исчезло бы с таба, оставшись непрочитанным на сервере.
                .onSuccess {
                    // Не ноль: `pending`-приглашения сервер считает непрочитанными,
                    // пока на них не ответили, — обнулив бейдж, мы бы разошлись
                    // с ним до следующего обновления (паритет с iOS).
                    unreadCount = _invites.value.count { it.status == InviteStatus.PENDING }
                    sessionStore.setUnreadNotifications(unreadCount)
                }
        }
    }

    fun acceptInvite(card: InviteCard) = actOnInvite(card) { repository.acceptInvite(card.roomId) }

    fun declineInvite(card: InviteCard) = actOnInvite(card) { repository.declineInvite(card.roomId) }

    /**
     * Выйти из группы прямо с карточки «вас добавили».
     *
     * Кнопка обязана быть здесь: человека добавили, не спросив, и если
     * единственное действие — «Открыть», отказаться можно только разыскав
     * настройки группы.
     */
    fun leaveFromCard(card: InviteCard) = actOnInvite(card) { repository.leaveRoom(card.roomId) }

    private fun actOnInvite(card: InviteCard, action: suspend () -> Unit) {
        viewModelScope.launch {
            try {
                action()
                _invites.value = _invites.value.filterNot { it.roomId == card.roomId }
                // Не только своя лента: принятое приглашение добавляет группу, а
                // выход убирает. Списки групп и друзей перезагружаются лишь по
                // dataVersion, а вкладки переживают переключение — без этого
                // «Группы» показывали бы старое до pull-to-refresh.
                sessionStore.noteDataChanged()
                reloadFirstPage()
            } catch (e: ApiException) {
                _errorMessage.value = humanErrorText(e)
            }
        }
    }

    private suspend fun reloadFirstPage() {
        try {
            val fetched = repository.notificationFeed(limit = PAGE_SIZE, offset = 0)
            val feed = fetched.value
            val page = feed.items
            generation++ // подгрузки, стартовавшие до этого момента, свой результат выбросят
            _invites.value = feed.invites
            seenThrough = feed.seenThrough
            unreadCount = feed.unreadCount
            sessionStore.setUnreadNotifications(feed.unreadCount)
            _state.value = UiState.Content(page)
            loadedCount = page.size
            // Из кеша листать некуда: следующая страница есть только на сервере,
            // и попытка её взять офлайн заканчивалась алертом «нет соединения»
            // поверх нормально показанной ленты (порт iOS ActivityViewModel).
            hasMore = !fetched.fromCache && page.size == PAGE_SIZE
            // Первый визит: экран открылся раньше, чем приехал ответ, и отметить
            // прочитанное до этой строки было нечем.
            if (isScreenVisible) markSeen()
        } catch (e: ApiException) {
            if (_state.value is UiState.Content) {
                _errorMessage.value = humanErrorText(e)
            } else {
                _state.value = UiState.Error(e.message)
            }
        }
    }

    private suspend fun loadMore() {
        if (!hasMore || _isLoadingMore.value) return
        val current = (_state.value as? UiState.Content)?.value ?: return
        val startedAt = generation
        _isLoadingMore.value = true
        try {
            val page = repository.activity(limit = PAGE_SIZE, offset = loadedCount).value
            if (startedAt != generation) return // лента перезагрузилась — страница устарела
            val base = (_state.value as? UiState.Content)?.value ?: current
            // Страховка от дублей при сдвиге offset (новые операции сверху).
            val known = base.mapTo(HashSet()) { it.operation.id }
            val fresh = page.filter { it.operation.id !in known }
            _state.value = UiState.Content(base + fresh)
            loadedCount += page.size
            // Страница целиком из дублей означает, что вниз двигаться некуда:
            // без этой отсечки цикл повторял бы один и тот же запрос.
            hasMore = page.size == PAGE_SIZE && fresh.isNotEmpty()
        } catch (e: ApiException) {
            _errorMessage.value = humanErrorText(e)
        } finally {
            _isLoadingMore.value = false
        }
    }
}
